File: low-level-design-java/src/inmemorycachedesign/dao/CacheDao.java
package inmemorycachedesign.dao;
import inmemorycachedesign.constants.EvictionPolicy;
import inmemorycachedesign.model.Cache;

import java.util.HashMap;
import java.util.LinkedList;


public class CacheDao {
    private static CacheDao cacheDao = null;
    private Cache cache=null;

    private CacheDao(){

    }

    public static CacheDao getInstance(){
        if(cacheDao == null){
            cacheDao=new CacheDao();
        }
        return cacheDao;
    }

    public void createCache(int size, EvictionPolicy evictionPolicy){
        cache = new Cache(size, evictionPolicy);
    }

    public Object get(Object obj){
        Object result = cache.getCacheStore().get(obj);

        // If object does not exit, then return null
        if(result == null){
            return null;
        }

        // If object exists, return result
        // update data, store, in accordance to Eviction policy
        cache.getCacheStore().remove(obj);
        cache.getDataStore().addFirst(obj);
        return result;
    }

    public void put(Object key, Object value){
        try{
            if(cache.getSize() == cache.getDataStore().size()){
                Object last = cache.getDataStore().removeLast();
                cache.getDataStore().remove(last);
            }
            //Remove key from LL if exit beforehand
            cache.getDataStore().remove(key);
            //Place it in front
            cache.getDataStore().addFirst(key);
            cache.getCacheStore().put(key, value);
        } catch (Exception e){
            System.out.println("Element is not added because of: " + e);
        }

    }

    public Boolean delete(Object object){
        if (cache.getDataStore().size() == 1) return true;

        try{
            cache.getDataStore().remove(object);
            cache.getCacheStore().remove(object);
            return true;
        }catch (Exception e){
            System.out.println("Delete function not working properly. " + e);
            return false;
        }
    }

    public boolean clear(){
        try {
            cache.setCacheStore(new HashMap<>());
            cache.setDataStore(new LinkedList<>());
            return true;
        }catch (Exception e){
            System.out.println("Cache is not clear because of " + e);
            return false;
        }
    }

    public void display(){
        for (Object object1: cache.getDataStore()){
            System.out.println(object1);
        }
    }


}

File: low-level-design-python/inmemorycachedesign/dao/CacheDao.py
import collections

from inmemorycachedesign.model.Cache import Cache


class CacheDao:
    __instance = None

    def __init__(self):
        self.cache = None

    def __new__(cls):
        if cls.__instance is None:
            cls.__instance = super(CacheDao, cls).__new__(cls)
        return cls.__instance

    def create_cache(self, size, eviction_policy):
        self.cache = Cache(size, eviction_policy)

    def get(self, key):
        result = self.cache.get_cache_store().get(key)
        if result is None:
            return None

        self.cache.get_cache_store().pop(key)
        self.cache.get_data_store().appendleft(key)
        return result

    def put(self, key, value):
        try:
            if self.cache.get_size() == len(self.cache.get_data_store()):
                last = self.cache.get_data_store().pop()
                if last in self.cache.get_data_store():
                    self.cache.get_data_store().pop(last)

            if key in self.cache.get_data_store():
                self.cache.get_data_store().remove(key)

            self.cache.get_data_store().appendleft(key)
            self.cache.get_cache_store()[key] = value
        except Exception as e:
            print("Element is not added because of: ", e)

    def delete(self, key):
        if len(self.cache.get_data_store()) == 1:
            return True

        try:
            self.cache.get_data_store().remove(key)
            self.cache.get_cache_store().pop(key)
            return True
        except Exception as e:
            print("Delete function not working properly: ", e)
            return False

    def clear(self):
        try:
            self.cache.set_caches_store(dict())
            self.cache.set_data_store(collections.deque)
            return True

        except Exception as e:
            print("Cache is not clear because of: ", e)
            return False

    def display(self):
        for i in self.cache.get_data_store():
            print(i)
